### PR TITLE
Added Prance

### LINKED ISSUE #116: Add support for loading schema with oneOf, anyOf, allOf, not $refs
Hi fellas,

I get this:

```
 File "/usr/local/lib/python3.8/site-packages/django_swagger_tester/testing.py", line 37, in validate_response
    verbose_error_message = format_response_tester_error(e, hint=e.response_hint)
  File "/usr/local/lib/python3.8/site-packages/django_swagger_tester/utils.py", line 47, in format_response_tester_error
    example_item = settings.loader_class.create_dict_from_schema(exception.schema)
  File "/usr/local/lib/python3.8/site-packages/django_swagger_tester/loaders.py", line 335, in create_dict_from_schema
    return self._iterate_schema_dict(schema)
  File "/usr/local/lib/python3.8/site-packages/django_swagger_tester/loaders.py", line 294, in _iterate_schema_dict
    elif read_type(value) == 'object':
  File "/usr/local/lib/python3.8/site-packages/django_swagger_tester/openapi.py", line 53, in read_type
    raise OpenAPISchemaError(
django_swagger_tester.exceptions.OpenAPISchemaError: Schema item has an invalid `type` attribute. The type should be a single string.
```

I then see this printed: 
`Schema item: {'allOf': [{'type': 'object', 'description': 'Read-only Product Serializer', 'properties': {'slug': {'type': 'string', 'readOnly': True, 'pattern': '^[-a-zA-Z0-9_]+$'}, 'loop': {'type': 'string', 'readOnly': True}, 'sites': {'type': 'array', 'items': {'type': 'object', 'description': 'read-only Site serializer', 'properties': {'id': {'type': 'integer', 'readOnly': True}, 'slug': {'type': 'string', 'readOnly': True, 'pattern': '^[-a-zA-Z0-9_]+$'}, 'domain': {'type': 'string', 'readOnly': True}}, 'required': ['domain', 'id', 'slug']}, 'readOnly': True}, 'acr': {'allOf': [{'enum': ['3', '4'], 'type': 'string'}], 'readOnly': True}}, 'required': ['acr', 'loop', 'sites', 'slug']}], 'readOnly': True}`

which I prettified:
```
{
  allOf: [
    {
      type: "object",
      description: "Read-only Product Serializer",
      properties: {
        slug: { type: "string", readOnly: True, pattern: "^[-a-zA-Z0-9_]+$" },
        loop: { type: "string", readOnly: True },
        sites: {
          type: "array",
          items: {
            type: "object",
            description: "read-only Site serializer",
            properties: {
              id: { type: "integer", readOnly: True },
              slug: {
                type: "string",
                readOnly: True,
                pattern: "^[-a-zA-Z0-9_]+$",
              },
              domain: { type: "string", readOnly: True },
            },
            required: ["domain", "id", "slug"],
          },
          readOnly: True,
        },
        acr: { allOf: [{ enum: ["3", "4"], type: "string" }], readOnly: True },
      },
      required: ["acr", "loop", "sites", "slug"],
    },
  ],
  readOnly: True,
};
```

I dont see any issue with the type key here.

### PR DESCRIPTION
**Pull request by Goldziher**
This PR adds prance to handle dereferencing.

It doesn't yet resolve #116, which will happen in a followup PR

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -45,15 +45,4 @@ ignore=
...
 
-exclude =
-    .git,
-    .idea,
-    __pycache__,
-    demo/*,
-    tests/*,
-    venv,
-    manage.py,
-    .venv
```
Comment (sondrelg) [inline .flake8:56]: Why did you remove this?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -3,52 +3,42 @@ repos:
...
       - id: flake8
-        additional_dependencies: [
-            'flake8-bugbear==20.1.4',  # Looks for likely bugs and design problems
-            'flake8-comprehensions==3.2.3',  # Looks for unnecessary generator functions that can be converted to list comprehensions
-            'flake8-deprecated==1.3',  # Looks for method deprecations
-            'flake8-use-fstring==1.1',  # Enforces use of f-strings over .format and %s
-            'flake8-print==3.1.4',  # Checks for print statements
-            'flake8-docstrings==1.5.0',  # Verifies that all functions/methods have docstrings
-            'flake8-type-annotations==0.1.0',  # Looks for misconfigured type annotations
-        ]
```
Comment (sondrelg) [inline .pre-commit-config.yaml:30]: Why did you remove these?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -3,52 +3,42 @@ repos:
...
       - id: pyupgrade
         args: ["--py36-plus"]
-  - repo: https://github.com/pycqa/isort
-    rev: 5.5.2
+  - repo: https://github.com/asottile/seed-isort-config
+    rev: v2.2.0
+    hooks:
+      - id: seed-isort-config
+  - repo: https://github.com/pre-commit/mirrors-isort
+    rev: "v5.7.0"
```
Comment (sondrelg) [inline .pre-commit-config.yaml:30]: Both of these repos say they're deprecated and that one should use `isort` directly. Unless you have a good reason to change this, I think it should keep the current config.

**Review by sondrelg — CHANGES_REQUESTED** [review]
Looks like there's almost 4000 changes here, where the strong majority is just single-to-double quotes. Please revert the quote-changes.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Goldziher) [reply .pre-commit-config.yaml:30]: for the typing stuff - you already had mypy in place, no need to have flake8 do an inferior job. For the rest of it - these did not actually contribute to find any errors in the code base. If we want these type of checks we better use pylint (although its somewhat dubious as well):

File: .pre-commit-config.yaml (hunk 3 above)
Comment (Goldziher) [reply .pre-commit-config.yaml]: done

File: .flake8 (hunk 1 above)
Comment (Goldziher) [reply .flake8:56]: because its not necessary. flake8 runs only against python files and i needed to run it against the stuff in tests.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Goldziher) [reply .pre-commit-config.yaml:30]: also the main reason - to enable running flake8 against the test files as well. This allowed me to find a lot of issues and fix them (especially missing and imports and unused variables etc). 

> Looks like there's almost 4000 changes here, where the strong majority is just single-to-double quotes. Please revert the quote-changes.

not possible, this is black.

Comment (sondrelg) [reply .pre-commit-config.yaml:30]: > not possible, this is black.

No, this happened because you removed the `double-quote-string-fixer` from the pre-commit config and ran it on the project.

> for the typing stuff - you already had mypy in place, no need to have flake8 do an inferior job

What does mypy have to do with:

- Fixing quotes
- Detecting private keys (I don't really care about this hook)
- Detecting bad list comprehensions
- Bugbear
- Method deprecations 
- Checking for print statements
- Checking for docstrings
- Checking that f-strings are used over format

The only flake8-hook related to mypy is the type annotations one, and mypy does not enforce type annotations. For example, this code would pass mypy's tests without errors

```python
def test(x, y):
    return x + y
    
test(1, 3)
```

With the flake8 type-annotations hook this would not pass the pre-commit check.

-------------------------

I think the best thing would be to undo these removals and run `pre-commit run --all-files`. That should eliminate the excess changes.

Let me know if you want me to do it :)

File: .flake8 (hunk 1 above)
Comment (sondrelg) [reply .flake8:56]: I have my virtual environments in my project directory, and venv's are propped full of python files. Think it would be better to only remove the `tests/*` line and keep everything else as it was.

Comment (Goldziher) [reply .flake8:56]: pre-commit doesnt run against git ignored files...

as you wish though.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Goldziher) [reply .pre-commit-config.yaml]: followed your suggestion. it worked!

File: response_tester/type_declarations.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    Exporter.export_anything(__name__)
+
+else:
+    """
+    if we are type checking, import/define the actual classes/types
+    We don't have to worry about circular imports here, since
+    the type checker does not actually run the code.
+    """
+    # noinspection PyUnresolvedReferences
+    from response_tester.loaders import BaseSchemaLoader, StaticSchemaLoader
```
Comment (sondrelg) [inline response_tester/type_declarations.py:43]: cool 😊

**Review by sondrelg — APPROVED** [review]
Nice! Pretty happy to merge this, but if this sounds OK, I think I will finalize #131, merge that, then merge this into master right after. 

Thinking I will sit down to do it tomorrow 🎉

**Comment by Goldziher** [comment]
> Nice! Pretty happy to merge this, but if this sounds OK, I think I will finalize #131, merge that, then merge this into master right after.
> 
> Thinking I will sit down to do it tomorrow 🎉

of course. I will continue working on my other PR when time allows - probably not before the weekend.
